### PR TITLE
Feature/inspections - open-ended inspections refactored

### PR DESCRIPTION
**Pull request by FritzViljoen**
Summary of this change:

Refactored inspections 

CHANGELOG entry:

### Added
- check_pallet(:failed_inspections)
- 
### Changed
- .
### Fixed
- .
### Removed
- services/failed_and_pending_pallet_inspections.rb
- check_pallet(:stock_grade)

How to test this code via the UI:

go path /finished_goods/inspection/inspections/inspection_results
select pallet ( still to be changed to accept scaner)
should list all applicable inspection_types for the pallet. 
add inspections 

PPECB is also handled with this improvement

### CODE REVIEW

File: lib/finished_goods/repositories/inspection_repo.rb [hunk 1]
```diff
@@ -21,98 +21,43 @@ class InspectionRepo < BaseRepo
...
-        match_tm = (inspection_type.applicable_tm_ids & hash[:tm_ids]).any?
-        match_tm_customer = (inspection_type.applicable_tm_customer_ids & hash[:tm_customer_ids]).any?
-        match_grade = (inspection_type.applicable_grade_ids & hash[:grade_ids]).any?
-        match_marketing_org_party_role = (inspection_type.applicable_marketing_org_party_role_ids & hash[:marketing_org_party_role_ids]).any?
-
-        if match_tm | match_grade | match_tm_customer | match_marketing_org_party_role
-          hash[:inspection_type_ids] << inspection_type_id
-          hash[:passed_default] << inspection_type.passed_default
-        end
+    def dataminer_report(yaml_file_name, for_grid: true, conditions: [])
```
Comment (jcsjcs) [inline lib/finished_goods/repositories/inspection_repo.rb:24]: This method should not be re-defined in the repo.
See comment under `passed_inspections`.

File: lib/finished_goods/repositories/inspection_repo.rb [hunk 2]
```diff
@@ -21,98 +21,43 @@ class InspectionRepo < BaseRepo
...
-                                       .get(:id))
-        .select_map(:inspection_type_id)
+      query = <<~SQL
+        SELECT
+          bool_and(passed) AS passed
+        FROM (#{inspection_results.runnable_sql}) inspection_results
+        WHERE active
+        GROUP BY inspection_results.pallet_id
+      SQL
+      DB[query].first&.dig(:passed)
```
Comment (jcsjcs) [inline lib/finished_goods/repositories/inspection_repo.rb:54]: Surely there is a more efficient query that can be written to just answer the question "has this pallet passed inspection?"

**Review by jcsjcs — CHANGES_REQUESTED** [review]

File: grid_definitions/dataminer_queries/inspection_types.yml [hunk 3]
```diff
@@ -1,44 +1,81 @@
...
+      (
+          SELECT STRING_AGG(fn_party_role_name(id), '; ')
+          FROM party_roles
+          WHERE id = ANY (coalesced.applicable_marketing_org_party_role_ids)
+      ) AS applicable_marketing_org_party_roles,
+      inspection_types.active,
+      inspection_types.created_at,
+      inspection_types.updated_at,
+      fn_current_status('inspection_types', inspection_types.id) AS status,
+      EXISTS(SELECT 1 FROM inspections WHERE inspection_type_id != inspection_types.id LIMIT 1) AS can_delete
```
[the remark this answers is not in the document]
Comment (FritzViljoen) [reply grid_definitions/dataminer_queries/inspection_types.yml:51]: Not exists reads better ill change it up

File: grid_definitions/dataminer_queries/inspection_types.yml [hunk 4]
```diff
@@ -363,6 +416,22 @@
     :group_avg: false
     :group_min: false
     :group_max: false
+  can_delete:
+    :name: can_delete
+    :sequence_no: 22
+    :caption: Can delete
+    :namespaced_name: 
+    :data_type: :string
```
[the remark this answers is not in the document]
Comment (FritzViljoen) [reply grid_definitions/dataminer_queries/inspection_types.yml:424]: cool

File: lib/finished_goods/repositories/inspection_repo.rb (hunk 1 above)
Comment (FritzViljoen) [reply lib/finished_goods/repositories/inspection_repo.rb:24]: ideally I would want this to be defined in the base_repo, and then have the base_interactor use the repo method.

File: lib/finished_goods/repositories/inspection_repo.rb (hunk 2 above)
Comment (FritzViljoen) [reply lib/finished_goods/repositories/inspection_repo.rb:54]: This is to have only one query to maintain

Comment (jcsjcs) [reply lib/finished_goods/repositories/inspection_repo.rb:54]: But it is incredibly inefficient - especially as it is called inside a loop where each iteration loads the YAML and creates a report object to get the SQL which queries far more data from the database than is required to answer this question.

The saving on maintenance and lack of a clear definition in the repo of what is "passed" is not worth it.

Comment (FritzViljoen) [reply lib/finished_goods/repositories/inspection_repo.rb:54]: We can change 

conditions: [{ col: 'inspection_results.pallet_id', op: '=', val: pallet_id }]

to use a list of pallet Ids

Comment (jcsjcs) [reply lib/finished_goods/repositories/inspection_repo.rb:54]: At this stage I would prefer this:
```sql
SELECT
  CASE WHEN inspection_results.inspection_type_code = 'PPECB' THEN
    pallets.inspected
    AND pallets.govt_inspection_passed
  WHEN inspections.active IS FALSE THEN
    NULL
  WHEN inspection_results.optional
    AND (inspections.inspector_id IS NULL) THEN
    TRUE
  WHEN (NOT inspection_results.optional)
    AND (inspections.inspector_id IS NULL) THEN
    inspection_results.passed_default
  ELSE
    inspections.passed
  END AS passed
FROM
  pallets
  LEFT JOIN (
    SELECT
      pallet_sequences.pallet_id,
      inspection_types.id AS inspection_type_id,
      inspection_types.inspection_type_code,
      inspection_types.optional,
      inspection_types.passed_default
    FROM
      pallet_sequences
      INNER JOIN inspection_types ON pallet_sequences.grade_id = ANY (inspection_types.applicable_grade_ids)
        OR pallet_sequences.marketing_org_party_role_id = ANY (inspection_types.applicable_marketing_org_party_role_ids)
        OR pallet_sequences.packed_tm_group_id = ANY (inspection_types.applicable_packed_tm_group_ids)
        OR pallet_sequences.target_customer_party_role_id = ANY (inspection_types.applicable_tm_customer_ids)
        OR pallet_sequences.target_market_id = ANY (inspection_types.applicable_tm_ids)
    WHERE
      pallet_sequences.pallet_id IS NOT NULL
    GROUP BY
      inspection_types.id,
      inspection_types.inspection_type_code,
      inspection_types.optional,
      inspection_types.passed_default,
      pallet_sequences.pallet_id) inspection_results ON inspection_results.pallet_id = pallets.id
  LEFT JOIN inspections ON inspections.pallet_id = inspection_results.pallet_id
    AND inspections.inspection_type_id = inspection_results.inspection_type_id
    AND inspections.active = TRUE
WHERE
  pallets.id = ?
```
And check with:
```ruby
DB[query, pallet_id].select_map.all?
```

**Pull request merged**